### ISSUE TITLE
Kryonet disconnects just after connecting

### ISSUE THREAD

**Issue by ghost** [EsotericSoftware/kryonet#35]
_From [cai...@gmail.com](https://code.google.com/u/101276589749535757218/) on July 07, 2013 19:02:06_

<b>What steps will reproduce the problem?</b>
1. start a server&#13;
2. start client&#13;
3. client keep sendTCP to server&#13;
4. server not send anything to client&#13;
&#13;
<b>What is the expected output? What do you see instead?</b>
[kryonet] Connection 1 connected: /127.0.0.1&#13;
[kryonet] Connection 1 disconnected.&#13;
in 12 seconds&#13;
&#13;
&#13;
<b>What version of the product are you using? On what operating system?</b>
2.20&#13;
&#13;
&#13;
<b>Please provide any additional information below.</b>
&#13;
Reason and how to fixed it:&#13;
&#13;
if i am keep writing lastWriteTime will bigger than the parameter "time" for method:&#13;
&#13;
com.esotericsoftware.kryonet.TcpConnection.needsKeepAlive(long time);&#13;
&#13;
so this will case "time - lastWriteTime" &lt; 0 and of course made it &lt; keepAliveMillis. That is why needsKeepAlive() will always return true, and client killed it by itself.&#13;
&#13;
i fixed by adding such code in the needsKeepAlive() method and isTimedOut (long time) method, you can see details in attachment.&#13;
&#13;
Hope this helpful!!&#13;
&#13;
BTW your frame is great! and simple to use!!&#13;
&#13;

**Attachment:** [compare.jpg](http://code.google.com/p/kryonet/issues/detail?id=34)

_Original issue: http://code.google.com/p/kryonet/issues/detail?id=34_

**Comment by breskeby** [comment]
hit the same issue. my problem was that I registered the UUID class which doesn't provide a public default constructor. replacing it solved the problem for me

**Comment by MarkSill** [comment]
@breskeby Exact same problem here. I'd thought I'd added default constructors to all registered classes, but there was one without. Adding a default constructor fixed it immediately.

**Comment by Isfirs** [comment]
I agree with @MarkSill, I faced the same problem yesterday and adding **ALL** default constructors fixed everything! Someone should spread this around.

**Comment by adudez96** [comment]
@Isfirs Same issue, same solution. Public default constructors need to be added to every single one.

**Comment by NathanSweet** [comment]
Sounds like two separate problems:

1. Not specifying both the same ports on client and server (TCP or TCP+UDP), and
2. Not providing default constructors.

What error does 2 provide? It should be explicit that a zero arg constructor is needed. Does it fail silently instead? If so, are you using StdInstantiatorStrategy? Using that is not recommended (it creates objects without calling _any_ constructor).

KryoNet is due for an update to the latest Kryo, that may help.

**Comment by Zhuinden** [comment]
My problem was multi-fold...

1.) initially I was using Kotlin data class, which doesn't create empty constructor out of the box

2.) Kotlin also wraps arrays and stuff with its own primitives, so I moved the `kryo.register` calls over to Java

3.) I didn't specify UDP port for the client even though I should have.

4.) I was trying to send TCP message back to the server using the `Connection` I received in the `connected` event callback, but apparently instead of that, I was supposed to use `client.sendTCP` instead of `connection.sendTCP`.

**Comment by CrashedLife** [comment]
> 

Old thread but I'd the same issue and this solved it.

**Comment by ghost** [comment]
> @breskeby Exact same problem here. I'd thought I'd added default constructors to all registered classes, but there was one without. Adding a default constructor fixed it immediately.

Had the same issue. This solved it immediately.
